### ISSUE TITLE
Add arm7l support so ROS2 can build on Pi Raspbian 

### ISSUE THREAD

**Issue by rydb** [ros2/mimick_vendor#11]
Every other ros2 package builds on raspian without issue but this one package is the only one that fails to build. It can't be pkg_ignored due to rcutils depending on it, and the ubuntu 20.04 pi image breaks pi peripherals and requires recompiling the kernel in order to use its GPIO properly. 

This is the error it throws when it tries to build:

```
--- stderr: mimick_vendor                      
Cloning into 'mimick-ext'...
Note: checking out 'feb30f7520d80f99344978901a4c8496e698004c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at feb30f7 Suppress cppcheck error for MMK_MANGLE (#12)
CMake Error at CMakeLists.txt:88 (message):
  Architecture 'armv7l' is not supported.


make[2]: *** [CMakeFiles/mimick-ext.dir/build.make:106: mimick-ext-prefix/src/mimick-ext-stamp/mimick-ext-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/mimick-ext.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

**Comment by hidmic** [comment]
If I'm not mistaken, `armv7l` is a 32 bits processor. So it's Mimick's arch detection logic that is too naive. @rydb would you mind tweaking this package's CMakeLists.txt, adding `-DCMAKE_GENERATOR_PLATFORM=arm32` [here](https://github.com/ros2/mimick_vendor/blob/06f4b3f3f98db1446c8751f27ff1b25bb73e4d8f/CMakeLists.txt#L67)? If it works locally for you, we can improve that detection logic.

**Comment by rydb** [comment]
> If I'm not mistaken, `armv7l` is a 32 bits processor. So it's Mimick's arch detection logic that is too naive. @rydb would you mind tweaking this package's CMakeLists.txt, adding `-DCMAKE_GENERATOR_PLATFORM=arm32` [here](https://github.com/ros2/mimick_vendor/blob/06f4b3f3f98db1446c8751f27ff1b25bb73e4d8f/CMakeLists.txt#L67)? If it works locally for you, we can improve that detection logic.

Actually, I looked at the mimick library and saw it built 64 bit arm processors and I remembered the pi 4 has a 64 bit processor. I'll set the pi to be 64 bit in its raspian configs and see if that allows it to build.

EDIT: Rest in pieces that idea. mimick vendor throws a low level error when I try setting it to 64 bit. I'll try your idea now

```
Starting >>> mimick_vendor
--- stderr: mimick_vendor
Cloning into 'mimick-ext'...
Note: checking out 'feb30f7520d80f99344978901a4c8496e698004c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at feb30f7 Suppress cppcheck error for MMK_MANGLE (#12)
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S: Assembler messages:
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:27: Error: ARM register expected -- `adr x16,.'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:29: Error: bad instruction `stp x7,x6,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:30: Error: bad instruction `stp x5,x4,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:31: Error: bad instruction `stp x3,x2,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:32: Error: bad instruction `stp x1,x0,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:34: Error: bad instruction `stp x16,x30,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:36: Error: ARM register expected -- `ldr x0,[x16,#-16]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:37: Error: ARM register expected -- `ldr x16,[x0,#8]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:38: Error: bad instruction `blr x16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:40: Error: bad instruction `ldp x16,x30,[sp]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:42: Error: ARM register expected -- `ldr x16,[x16,#-16]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:43: Error: ARM register expected -- `ldr x16,[x16]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:44: Error: bad instruction `blr x16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:45: Error: ARM register expected -- `cmp x0,#0'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:46: Error: unexpected character `n' in type specifier
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:46: Error: bad instruction `b.ne ret_ctx'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:48: Error: bad instruction `ldp x16,x30,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:49: Error: bad instruction `ldp x1,x0,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:50: Error: bad instruction `ldp x3,x2,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:51: Error: bad instruction `ldp x5,x4,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:52: Error: bad instruction `ldp x7,x6,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:54: Error: ARM register expected -- `ldr x16,[x16,#-8]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:55: Error: bad instruction `br x16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:59: Error: bad instruction `ldp x16,x30,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:60: Error: bad instruction `ldp x1,x0,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:61: Error: bad instruction `ldp x3,x2,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:62: Error: bad instruction `ldp x5,x4,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:63: Error: bad instruction `ldp x7,x6,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:64: Error: bad instruction `stp x16,x30,[sp,#-16]!'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:66: Error: ARM register expected -- `ldr x16,[x16,#-16]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:67: Error: ARM register expected -- `ldr x16,[x16,#16]'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:68: Error: bad instruction `blr x16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:70: Error: bad instruction `ldp x16,x30,[sp],#16'
/home/pi/ros2_foxy/build/mimick_vendor/mimick-ext-prefix/src/mimick-ext/src/asm/trampoline-aarch64.S:71: Error: bad instruction `ret'
make[5]: *** [CMakeFiles/mimick.dir/build.make:63: CMakeFiles/mimick.dir/src/asm/trampoline-aarch64.S.o] Error 1
make[4]: *** [CMakeFiles/Makefile2:112: CMakeFiles/mimick.dir/all] Error 2
make[3]: *** [Makefile:141: all] Error 2
make[2]: *** [CMakeFiles/mimick-ext.dir/build.make:111: mimick-ext-prefix/src/mimick-ext-stamp/mimick-ext-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/mimick-ext.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
Failed   <<< mimick_vendor [22.0s, exited with code 2]
```

**Comment by rydb** [comment]
> If I'm not mistaken, `armv7l` is a 32 bits processor. So it's Mimick's arch detection logic that is too naive. @rydb would you mind tweaking this package's CMakeLists.txt, adding `-DCMAKE_GENERATOR_PLATFORM=arm32` [here](https://github.com/ros2/mimick_vendor/blob/06f4b3f3f98db1446c8751f27ff1b25bb73e4d8f/CMakeLists.txt#L67)? If it works locally for you, we can improve that detection logic.

Tried the idea, but now its throwing a different but similar error, this time about a make file.


```
Starting >>> mimick_vendor
--- stderr: mimick_vendor
Cloning into 'mimick-ext'...
Note: checking out 'feb30f7520d80f99344978901a4c8496e698004c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at feb30f7 Suppress cppcheck error for MMK_MANGLE (#12)
CMake Error at CMakeLists.txt:7 (project):
  Generator

    Unix Makefiles

  does not support platform specification, but platform

    arm32

  was specified.


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
make[2]: *** [CMakeFiles/mimick-ext.dir/build.make:106: mimick-ext-prefix/src/mimick-ext-stamp/mimick-ext-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/mimick-ext.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
Failed   <<< mimick_vendor [28.2s, exited with code 2]
```

**Comment by hidmic** [comment]
I was hoping we could force the detected arch for simplicity, but it won't budge. I'll take a look.

FYI @rydb, we're vendoring https://github.com/ros2/Mimick, not https://github.com/Snaipe/Mimick. I haven't found the time to submit our changes upstream.

**Comment by rydb** [comment]
> I was hoping we could force the detected arch for simplicity, but it won't budge. I'll take a look.
> 
> FYI @rydb, we're vendoring https://github.com/ros2/Mimick, not https://github.com/Snaipe/Mimick. I haven't found the time to submit our changes upstream.

How do I find the Mimick library ros2 uses when building mimick_vendor from source so I can try the work around mentioned in https://github.com/Snaipe/Mimick/issues/17#issue-733092525 ? I can't find the ros2 version of Mimick in the "vcs import src < ros2.repos" files.

**Comment by goekce** [comment]
FYI: I had also the same issue and created a the PR https://github.com/Snaipe/Mimick/pull/15. But https://github.com/Snaipe/Mimick/issues/17#issue-733092525 looks better to me.

**Comment by rydb** [comment]
I solved the issue by using Snaipe/Mimick#17 (comment) 's workaround. I made a pull request to have it added to ros2 Mimick.

https://github.com/ros2/Mimick/pull/14
